File: algorithms/bfs/typescript/bfs.ts
export function bfs(graph: ReadonlyMap<number, readonly number[]>, start: number): number[] {
  const visited = new Set<number>()
  const queue: number[] = [start]
  let dequeueIndex = 0
  const result: number[] = []

  visited.add(start)

  while (dequeueIndex < queue.length) {
    const node = queue[dequeueIndex++]
    result.push(node)

    const neighbors = graph.get(node) ?? []
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor)
        queue.push(neighbor)
      }
    }
  }

  return result
}

File: algorithms/bfs/go/bfs.go
package main

// graph は読み取り専用で使用する（Go に ReadonlyMap 相当はないため、変更しない前提）。
func BFS(graph map[int][]int, start int) []int {
	visited := map[int]bool{start: true}
	queue := []int{start}
	dequeueIndex := 0
	result := []int{}

	for dequeueIndex < len(queue) {
		node := queue[dequeueIndex]
		dequeueIndex++
		result = append(result, node)

		for _, neighbor := range graph[node] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return result
}
